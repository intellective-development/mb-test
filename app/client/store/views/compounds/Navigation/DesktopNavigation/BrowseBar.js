// @flow

import * as React from 'react';
import _, { sortBy } from 'lodash';
import bindClassNames from 'shared/utils/bind_classnames';
import type { ContentLayout } from 'store/business/content_layout';
import { content_module_helpers } from 'store/business/content_module';
import type { ListLink } from 'store/business/content_module';
import connectToObservables from 'shared/components/higher_order/connect_observables';
import locationStream from 'legacy_store/router/location_stream';

import { NAVIGATION_CATEGORY_DEFAULTS } from '../constants';
import withContentLayout from '../../../compounds/ContentLayout';
import SearchInput from './SearchInput';
import CategoryDropdown from './CategoryDropdown';
import { MBLink, withUniqueId } from '../../../elements';

import styles from './BrowseBar.scss';

const cn = bindClassNames(styles);

const NAVIGATION_CATEGORY_SECTION_ID = 'category_list';

type BrowseBarProps = {|
  show: boolean,
  location_fragment: string,
  content_layout: ContentLayout
|};
type BrowseBarState = {|
  selected_category_name: ?string,
  dropdown_open: boolean
|};

class BrowseBar extends React.Component<BrowseBarProps, BrowseBarState> {
  state = { selected_category_name: null, dropdown_open: false }

  componentWillReceiveProps(next_props: BrowseBarProps){
    if (this.props.location_fragment !== next_props.location_fragment){
      this.hideDropdown();
    }
  }

  showDropdown = (category_name: string) => {
    this.setState({selected_category_name: category_name, dropdown_open: true});
  }

  hideDropdown = () => {
    // only changing dropdown_open and not selected_category_name allows the dropdown to animate out
    this.setState({dropdown_open: false});
  }

  render(){
    const { content_layout } = this.props;
    const { selected_category_name, dropdown_open } = this.state;

    let category_links = NAVIGATION_CATEGORY_DEFAULTS;
    let selected_link = {};
    let dropdown_content_modules = [];
    if (content_layout){
      const content_module_sections = content_module_helpers.groupContentModulesBySection(content_layout.content);

      category_links = _.get(content_module_sections, `${NAVIGATION_CATEGORY_SECTION_ID}[0].config.links`) || [];
      category_links = sortBy(
        category_links,
        [
          ({ name }) =>
            ['Wine', 'Liquor', 'Beer', 'Mixers', 'Gifts'].indexOf(name)
        ]
      );
      selected_link = category_links.find(link => link.internal_name === selected_category_name) || {};
      dropdown_content_modules = content_module_sections[selected_link.dropdown_section_id] || [];
    }

    const dropdown_visible = dropdown_open && !!selected_category_name && !_.isEmpty(dropdown_content_modules);

    return (
      <nav className={styles.cmDBrowseBar}>
        <div onMouseLeave={this.hideDropdown}>
          <ul className={styles.cmDBrowseBar_CategoryList}>
            {category_links.map(link => (
              <CategoryElement
                key={link.internal_name}
                link={link}
                selected={(link.internal_name === selected_category_name) && dropdown_open}
                selectNavigationCategory={this.showDropdown} />
            ))}
          </ul>
          <CategoryDropdown
            is_available={!!(selected_link && selected_link.action_url)}
            selected_category_name={selected_category_name}
            content_modules={dropdown_content_modules}
            is_visible={dropdown_visible} />
        </div>
        <div className={styles.cmDBrowseBar_Search_Spacer} />
        <SearchInput />
      </nav>
    );
  }
}

type CategoryElementProps = {|
  link: ListLink,
  selected: boolean,
  selectNavigationCategory: (string) => void
|}
class CategoryElement extends React.Component<CategoryElementProps> {
  handleMouseEnter = () => {
    this.props.selectNavigationCategory(this.props.link.internal_name);
  }

  render(){
    const { link, selected } = this.props;
    const { name, action_url } = link;
    const disabled = !action_url;

    return (
      <li onMouseEnter={this.handleMouseEnter}>
        <MBLink.Text
          href={action_url}
          disabled={disabled}
          standard={false}
          className={cn('cmDBrowseBar_CategoryLink', {
            cmDBrowseBar_CategoryLink__Selected: selected,
            cmDBrowseBar_CategoryLink__Disabled: disabled
          })}>
          {name}
        </MBLink.Text>
      </li>
    );
  }
}

const BrowseBarContentLayout = withUniqueId('content_layout_id')(withContentLayout('Web_Navigation_Desktop_Category')(BrowseBar));
export default connectToObservables(BrowseBarContentLayout, {
  location_fragment: locationStream.map((location) => location.fragment)
});
export const __private__ = {
  BrowseBar
};

