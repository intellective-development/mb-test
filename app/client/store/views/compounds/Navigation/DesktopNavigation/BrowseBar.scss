@import "~client/store/views/style/variables";
@import "./_variables";

$browse-bar-height: 54px;
.cmDBrowseBar {
  border-top: 1px solid $brand-light-grey;
  height: $browse-bar-height;
  display: flex;
  align-items: center;
}

.cmDBrowseBar_CategoryList {
  display: flex;
  flex-grow: 1;
  margin: 0; // TODO: should this be on an element like MBList?
}

.cmDBrowseBar_CategoryLink {
  @include font-size(13);
  display: inline-block;
  height: $browse-bar-height;

  text-transform: uppercase;
  font-weight: $heavy-font-weight;
  letter-spacing: 2px;
  color: $brand-black;
  line-height: $browse-bar-height; // center vertically
  padding: 0 28px;
  padding-top: 2px;
  border-bottom: 4px solid transparent;

  transition: border-color 150ms ease-in-out;

  &:hover,
  &.cmDBrowseBar_CategoryLink__Selected, {
    border-bottom-color: $brand-red;
  }

  @media only screen and (max-width: $nav-contract-max-width) {
    padding: 0 15px;
  }
}

.cmDBrowseBar_CategoryLink__Disabled{
  color: $brand-medium-grey;
}

// Search Box

$search-icon-spacing: 50px;

.cmDBrowseBar_Search_Spacer {
  flex: 1;
}

.cmDBrowseBar_Search_Container {
  position: relative;
  max-width: 372px;
  flex: auto;
  display: flex;
  justify-content: flex-end;

  @media only screen and (max-width: $nav-contract-max-width) {
    padding-right: $standard-gutter;
  }
}

.cmDBrowseBar_Search_Wrapper {
  position: relative;
  flex: auto;
  max-width: 372px;
}

.cmDBrowseBar_Search_Icon {
  position: absolute;
  top: 3px;
  left: 13px;

  // ensures it won't block hover/focus events for the input it is sitting on top of
  pointer-events: none;
}

input.cmDBrowseBar_Search_Input {
  @include font-size(13);
  -webkit-appearance: none;
  background: $brand-lighter-grey;
  border-radius: 4px;
  border: 1px solid $brand-lighter-grey; // match the border color to the background for non-focused
  color: $brand-black;
  height: 32px;
  outline: none;
  width: 100%;
  padding: 0; // ensure input is usable in firefox (by overriding foundation)
  padding-left: $search-icon-spacing;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-transform: capitalize;

  &:hover{
    border-color: $brand-light-grey;
  }
  &:focus {
    border-color: $brand-medium-grey;
  }
  &::placeholder {
    text-transform: initial;
    color: $brand-medium-grey;
  }
}
