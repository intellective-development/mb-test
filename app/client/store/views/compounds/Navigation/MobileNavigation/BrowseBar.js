// @flow
import * as React from 'react';
import bindClassNames from 'shared/utils/bind_classnames';
import connectToObservables from 'shared/components/higher_order/connect_observables';
import locationStream from 'legacy_store/router/location_stream';


import { MBClickOutside, MBDynamicIcon, MBIcon, MBInput, MBText, MBTouchable } from '../../../elements';
import Search from '../shared/Search';
import CategoryDropdown from './CategoryDropdown';
import styles from './BrowseBar.scss';

const cn = bindClassNames(styles);

type BrowseBarProps = {|
  location_fragment: string
|};
type BrowseBarState = {|
  search_active: boolean,
  category_dropdown_open: boolean
|};
class BrowseBar extends React.Component<BrowseBarProps, BrowseBarState> {
  state = {search_active: false, category_dropdown_open: false}

  componentWillReceiveProps(next_props: BrowseBarProps){
    if (this.props.location_fragment !== next_props.location_fragment){
      this.hideCategoryDropdown();
    }
  }

  openSearch = () => {
    this.setState({search_active: true});
  }

  closeSearch = () => {
    this.setState({search_active: false});
  }

  toggleCategoryDropdown = () => {
    this.setState(prev_state => ({category_dropdown_open: !prev_state.category_dropdown_open}));
  }

  hideCategoryDropdown = () => {
    this.setState({category_dropdown_open: false});
  }

  render(){
    const { search_active, category_dropdown_open } = this.state;

    return (
      <nav className={styles.cmMBrowseBar}>
        <Categories
          search_active={search_active}
          category_dropdown_open={category_dropdown_open}
          toggleCategoryDropdown={this.toggleCategoryDropdown}
          hideCategoryDropdown={this.hideCategoryDropdown} />
        <SearchInput
          openSearch={this.openSearch}
          closeSearch={this.closeSearch}
          search_active={search_active} />
      </nav>
    );
  }
}

const Categories = ({search_active, category_dropdown_open, toggleCategoryDropdown, hideCategoryDropdown}) => {
  return (
    <MBClickOutside handleClickOutside={hideCategoryDropdown} disableOnClickOutside={!category_dropdown_open}>
      <div className={cn('cmMBrowseBar_LeftWrapper', {cmMBrowseBar_LeftWrapper__SearchActive: search_active})}>
        <MBTouchable
          className={cn('cmMBrowseBar_Left', {cmMBrowseBar_Left__Active: category_dropdown_open})}
          onClick={toggleCategoryDropdown}>
          <MBText.Span className={styles.cmMBrowseBar_CategoryPrompt} >Categories</MBText.Span>
          <MBIcon name="down_arrow_red" className={cn('cmMBrowseBar_DicloseIcon', {cmMBrowseBar_DicloseIcon__Active: category_dropdown_open})} />
        </MBTouchable>
        <CategoryDropdown is_showing={category_dropdown_open} hide={hideCategoryDropdown} />
      </div>
    </MBClickOutside>
  );
};

type SearchInputProps = {
  search_active: boolean,
  openSearch: () => void,
  closeSearch: () => void
}

const SearchInput = ({search_active, openSearch, closeSearch}: SearchInputProps) => {
  return (
    <Search
      onClose={closeSearch}
      renderInput={({ value, clear, onKeyDown, onChange, onFocus, onBlur}) => (
        <div className={styles.cmMBrowseBar_Right}>
          <div className={styles.cmMBrowseBar_Search_Wrapper}>
            <MBDynamicIcon
              name="search"
              width={29}
              height={29}
              className={cn('cmMBrowseBar_SearchIcon', {cmMBrowseBar_SearchIcon__SearchActive: search_active})} />
            <MBInput.Input
              value={value}
              onChange={onChange}
              onKeyDown={onKeyDown}
              onFocus={(event) => {
                openSearch();
                onFocus(event);
              }}
              className={styles.cmMBrowseBar_SearchInput}
              placeholder="Search products" />
          </div>
          <MBTouchable
            onClick={clear}
            className={cn('cmMBrowseBar_ClearSearch', {
              cmMBrowseBar_ClearSearch__SearchEmpty: !value,
              cmMBrowseBar_ClearSearch__SearchActive: !search_active
            })}>
            <MBIcon name="clear" />
          </MBTouchable>
          <MBTouchable onClick={() => { onBlur(); closeSearch(); }}>
            <MBText.Span className={cn('cmMBrowseBar_CancelSearch', {cmMBrowseBar_CancelSearch__SearchActive: search_active})}>
              Cancel
            </MBText.Span>
          </MBTouchable>
        </div>
      )} />
  );
};

export default connectToObservables(BrowseBar, {
  location_fragment: locationStream.map((location) => location.fragment)
});
