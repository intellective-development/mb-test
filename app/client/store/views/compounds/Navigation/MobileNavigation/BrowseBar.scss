@import "~client/store/views/style/variables";

$browse-bar-height: 34px;
$search-disclose-animation-duration: 200ms;

.cmMBrowseBar {
  display: flex;
  height: $browse-bar-height;
}

.cmMBrowseBar_LeftWrapper {
  display: flex;
  flex: none;

  // animation
  transform: translateZ(0); // iOS performance hack
  transition: all $search-disclose-animation-duration ease-in-out;
  max-width: 150px; // roughly the right size
  opacity: 1;
}
.cmMBrowseBar_LeftWrapper__SearchActive {
  max-width: 0;
  opacity: 0;
}

.cmMBrowseBar_Left {
  transform: translateZ(0); // iOS performance hack
  display: flex;
  align-items: center;
  padding-right: $standard-gutter;
  border-right: $default-border;
  padding-left: $standard-gutter;
  transition: all 250ms ease-in-out;
}
.cmMBrowseBar_Left__Active {
  background-color: $brand-pale-blue;
}

.cmMBrowseBar_Right {
  display: flex;
  align-items: center;
  flex: auto;
  padding-right: $standard-gutter;
  margin-left: 12px;
  overflow-x: hidden;
}

.cmMBrowseBar_CategoryPrompt {
  font-weight: $heavy-font-weight;
  margin-top: 1px;
}
.cmMBrowseBar_DicloseIcon {
  margin-left: 7px;
  display: inline-block;
  transition: transform 120ms ease-in-out;
}
.cmMBrowseBar_DicloseIcon__Active { transform: rotate(180deg); }


$search-icon-spacing: 36px;

.cmMBrowseBar_Search_Wrapper {
  display: flex;
  position: relative;
  flex: auto;
}
.cmMBrowseBar_SearchIcon {
  position: absolute;
  left: 0;
  top: 3px;

  // ensures it won't block hover/focus events for the input it is sitting on top of
  pointer-events: none;
}

input.cmMBrowseBar_SearchInput { // input used to override system styles
  @include font-size(14);
  flex: auto;
  height: $browse-bar-height;
  color: $brand-black;
  border: none;
  padding: 0;
  color: $brand-black;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-transform: capitalize;
  padding-left: $search-icon-spacing;

  &::placeholder {
    text-transform: initial;
    color: $brand-medium-grey;
  }
}

.cmMBrowseBar_ClearSearch {
  cursor: pointer;
  backface-visibility: hidden; // prevents clipping with opacity
  padding: 0 8px 2px;
  margin-right: 8px;
  transition: all 120ms ease-in-out;
  opacity: 0.75;

  &:hover {
    opacity: 1;
  }
}
.cmMBrowseBar_ClearSearch__SearchEmpty {
  visibility: hidden;
  opacity: 0;
}
.cmMBrowseBar_ClearSearch__SearchActive {
  display: none;
}
.cmMBrowseBar_CancelSearch {
  @include font-size(14);

  font-weight: $heavy-font-weight;
  color: $brand-red;
  display: block;
  transition: all $search-disclose-animation-duration ease-in-out;
  transform: translateY(5px);
  opacity: 0;
  padding-top: 2px;
}
.cmMBrowseBar_CancelSearch__SearchActive {
  opacity: 1;
  transform: translateY(0);
}
