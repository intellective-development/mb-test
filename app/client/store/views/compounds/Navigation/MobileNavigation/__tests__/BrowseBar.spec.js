
import React from 'react';

import TestProvider from 'store/views/__tests__/utils/TestProvider';
import BrowseBar from '../BrowseBar';

describe('BrowseBar', () => {
  it('renders', () => {
    expect(render(
      <TestProvider>
        <BrowseBar />
      </TestProvider>
    )).toMatchSnapshot();
  });
});
