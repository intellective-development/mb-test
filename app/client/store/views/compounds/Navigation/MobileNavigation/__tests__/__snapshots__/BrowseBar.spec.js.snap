// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowseBar renders 1`] = `
<nav>
  <div
    class="cmMBrowseBar_LeftWrapper"
  >
    <div
      class="el-touchable cmMBrowseBar_Left"
      role="button"
      tabindex="-1"
    >
      <span
        class="elMBText elMBText_ResetSpacing"
      >
        Categories
      </span>
      <img
        alt="down_arrow_red"
        class="el-mbicon--down_arrow_red cmMBrowseBar_DicloseIcon"
        src="/assets/components/elements/mb-icon/down_arrow_red.png"
        srcset="/assets/components/elements/mb-icon/down_arrow_red@2x.png 2x, /assets/components/elements/mb-icon/down_arrow_red@3x.png 3x"
      />
    </div>
    <div
      class="cmMCategoryDropdown_Wrapper"
    >
      <ul
        class="cmMCategoryDropdown"
      >
        <li>
          <a
            class="cmMCategoryDropdown_Element"
            href="/store/category/wine"
          >
            <span
              class="elMBText elMBText_ResetSpacing"
            >
              wine
            </span>
          </a>
        </li>
        <li>
          <a
            class="cmMCategoryDropdown_Element"
            href="/store/category/liquor"
          >
            <span
              class="elMBText elMBText_ResetSpacing"
            >
              liquor
            </span>
          </a>
        </li>
        <li>
          <a
            class="cmMCategoryDropdown_Element"
            href="/store/category/beer"
          >
            <span
              class="elMBText elMBText_ResetSpacing"
            >
              beer
            </span>
          </a>
        </li>
        <li>
          <a
            class="cmMCategoryDropdown_Element"
            href="/store/category/mixers"
          >
            <span
              class="elMBText elMBText_ResetSpacing"
            >
              mixers
            </span>
          </a>
        </li>
        <li>
          <a
            class="cmMCategoryDropdown_Element"
            href="/store/promos/gift-guide"
          >
            <span
              class="elMBText elMBText_ResetSpacing"
            >
              gifts
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="cmSearch__ClickOutsideTarget"
  >
    <div>
      <div>
        <div
          class="elMBDynamicIcon cmMBrowseBar_SearchIcon"
        >
          <svg
            class=""
            height="29"
            viewBox="0 0 32 32"
            width="29"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="#222"
              fill-rule="nonzero"
            >
              <path
                d="M14.81 24.342c-5.145 0-9.31-4.22-9.31-9.42C5.5 9.72 9.665 5.5 14.81 5.5c5.145 0 9.31 4.22 9.31 9.421 0 5.2-4.165 9.421-9.31 9.421zm0-2c4.035 0 7.31-3.32 7.31-7.42 0-4.102-3.275-7.422-7.31-7.422-4.034 0-7.31 3.32-7.31 7.421 0 4.102 3.276 7.421 7.31 7.421z"
              />
              <path
                d="M20.251 14.921h-2c0-1.936-1.543-3.5-3.44-3.5-.833 0-1.617.3-2.237.84l-1.313-1.509a5.388 5.388 0 0 1 3.55-1.33c3.007 0 5.44 2.464 5.44 5.5zM20.331 21.94l1.424-1.406 4.499 4.56-1.424 1.404z"
              />
            </g>
          </svg>
        </div>
        <input
          class="elMBInput"
          placeholder="Search products"
          value=""
        />
      </div>
      <div
        class="el-touchable cmMBrowseBar_ClearSearch cmMBrowseBar_ClearSearch__SearchEmpty cmMBrowseBar_ClearSearch__SearchActive"
        role="button"
        tabindex="-1"
      >
        <img
          alt="clear"
          class="el-mbicon--clear"
          src="/assets/components/elements/mb-icon/clear.png"
          srcset="/assets/components/elements/mb-icon/clear@2x.png 2x, /assets/components/elements/mb-icon/clear@3x.png 3x"
        />
      </div>
      <div
        class="el-touchable"
        role="button"
        tabindex="-1"
      >
        <span
          class="elMBText elMBText_ResetSpacing cmMBrowseBar_CancelSearch"
        >
          Cancel
        </span>
      </div>
    </div>
    <div
      class="cmAutocompleteDropdown"
    />
  </div>
</nav>
`;
